lib/lsblk/pairs.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2018 Balena.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Drive, Mountpoint } from '..';
+
+interface Dict<T> {
+	[K: string]: T;
+}
+
+function parseLsblkLine(line: string): Dict<string> {
+	const data: Dict<string> = {};
+	let offset = 0;
+	let key = '';
+	let value = '';
+
+	const keyChar = /[^"=]/;
+	const whitespace = /\s+/;
+	const escape = '\\';
+	let state = 'key';
+
+	while (offset < line.length) {
+		if (state === 'key') {
+			while (keyChar.test(line[offset])) {
+				key += line[offset];
+				offset += 1;
+			}
+			if (line[offset] === '=') {
+				state = 'value';
+				offset += 1;
+			}
+		} else if (state === 'value') {
+			if (line[offset] !== '"') {
+				throw new Error(`Expected '"', saw "${line[offset]}"`);
+			}
+			offset += 1;
+			while (line[offset] !== '"' && line[offset - 1] !== escape) {
+				value += line[offset];
+				offset += 1;
+			}
+			if (line[offset] !== '"') {
+				throw new Error(`Expected '"', saw "${line[offset]}"`);
+			}
+			offset += 1;
+			data[key.toLowerCase()] = value.trim();
+			key = '';
+			value = '';
+			state = 'space';
+		} else if (state === 'space') {
+			while (whitespace.test(line[offset])) {
+				offset += 1;
+			}
+			state = 'key';
+		} else {
+			throw new Error(`Undefined state "${state}"`);
+		}
+	}
+
+	return data;
+}
+
+function parseLsblk(output: string): Dict<string>[] {
+	return output
+		.trim()
+		.split(/\r?\n/g)
+		.map(parseLsblkLine);
+}
+
+function consolidate(
+	devices: Dict<string>[],
+): (Dict<string> & { mountpoints: Mountpoint[] })[] {
+	const primaries = devices.filter(device => {
+		return (
+			device.type === 'disk' &&
+			!device.name.startsWith('ram') &&
+			!device.name.startsWith('sr')
+		);
+	});
+
+	return primaries.map(device => {
+		return Object.assign({}, device, {
```

Comment:
I'd like to but typescript complains about:
```
Type '{ mountpoints: Mountpoint[]; }' is not assignable to type 'Dict<string> & { mountpoints: Mountpoint[]; }'.
  Type '{ mountpoints: Mountpoint[]; }' is not assignable to type 'Dict<string>'.
    Property 'mountpoints' is incompatible with index signature.
      Type 'Mountpoint[]' is not assignable to type 'string'.

 92  	return {
     	~~~~~~~~
 93  		...device,
    ~~~~~~~~~~~~~
... 
109  			),
    ~~~~~~
110  	}
    ~~~
```

Reply:
Urg, yeah that looks like it might be a bug, I swear `Object.assign` and spread operators should behave the same.  I expect it's to do with the index type